runtime/src/debug/cpp/SourceInfo.cpp

```diff
@@ -92,28 +106,38 @@ bool TryInitializeCoreSymbolication() {
 } // namespace
 
 extern "C" struct SourceInfo Kotlin_getSourceInfo(void* addr) {
-  SourceInfo result = { .fileName = nullptr, .lineNumber = -1, .column = -1 };
+  __block SourceInfo result = { .fileName = nullptr, .lineNumber = -1, .column = -1 };
 
   static bool csIsAvailable = TryInitializeCoreSymbolication();
 
   if (csIsAvailable) {
     unsigned long long address = static_cast<unsigned long long>((uintptr_t)addr);
 
     CSSymbolOwnerRef symbolOwner = CSSymbolicatorGetSymbolOwnerWithAddressAtTime(symbolicator, address, kCSNow);
-    CSSourceInfoRef sourceInfo = CSSymbolOwnerGetSourceInfoWithAddress(symbolOwner, address);
-    if (!CSIsNull(sourceInfo)) {
-      const char* fileName = CSSourceInfoGetPath(sourceInfo);
-      if (fileName != nullptr) {
-        result.fileName = fileName;
-        uint32_t lineNumber = CSSourceInfoGetLineNumber(sourceInfo);
-        if (lineNumber != 0) {
-          result.lineNumber = lineNumber;
-          result.column = CSSourceInfoGetColumn(sourceInfo);
-        }
-      }
-    }
+    if (CSIsNull(symbolOwner))
+      return result;
+    CSSymbolRef symbol = CSSymbolOwnerGetSymbolWithAddress(symbolOwner, address);
+    if (CSIsNull(symbol))
+      return result;
+
+    CSSymbolForeachSourceInfo(symbol,
+      ^(CSSourceInfoRef ref) {
+          uint32_t lineNumber = CSSourceInfoGetLineNumber(ref);
+          CSRange range = CSSourceInfoGetRange(ref);
+          if (lineNumber != 0
+              && address >= range.location
+              && address < range.location + range.length) {
+            RuntimeAssert(result.fileName == nullptr, "shouldn't be more than one value per address");
```

Comment:
What does guarantee this?
When `CSSymbolRef` has multiple `CSSourceInfoRef`s, what do they look like?

Reply:
I've seen several SCSourceInfoRefs per CSSymbolRef on inlines. Current source info selection based on range, but we I haven't got enough information about CoreSymbolisation if it enough.

Reply:
Can we rely on the order of `SCSourceInfoRef`s here?

Reply:
Looks so, but it also iterates source info of closing scopes too.

Reply:
I don't get you.
Can we always use first or last `CSSourceInfoRef`?

Reply:
no, more likely [length/2]

Reply:
I see.
Let's stop reverse engineering for now then :)
Current heuristic is probably ok (although I have no idea what it means), but let's remove the assertion (since we don't now if the heuristic checks are enough).

Reply:
ok